broadway/api/handlers/stream.py

```diff
@@ -0,0 +1,38 @@
+from tornado import gen, web
+from tornado.iostream import StreamClosedError
+
+from broadway.api.handlers.base import BaseAPIHandler
+from broadway.api.decorators.auth import authenticate_course
+
+
+class GradingJobStreamHandler(BaseAPIHandler):
+    def initialize(self):
+        # Prepare for Server-Sent Events
+        self.set_header("Content-Type", "text/event-stream")
+        self.set_header("Cache-Control", "no-cache")
+
+        self._id = id(self)
+
+    @gen.coroutine
+    def publish(self, event, data):
+        try:
+            self.write(f"event: {event}" f"data: {data}\n\n")
+            yield self.flush()
```

Comment:
If the event is `GradingRunState.FAILED` or `GradingRunState.FINISHED`, there won't be any more events sent to this job. In that case, we should unregister the stream queue and close the stream (to prevent the stream queue from hogging memory forever).

Reply:
I'm sending a sentinel value to each queue now. The stream will pop it from the queue, recognize the value, and unregister itself.

Reply:
Maybe we can pass data in a json format for easier parsing on the client side?